lib/change_stream.js

```diff
@@ -25,48 +31,73 @@ var cursorOptionNames = ['maxAwaitTimeMS', 'collation', 'readPreference'];
  * @fires ChangeStream#error
  * @return {ChangeStream} a ChangeStream instance.
  */
-var ChangeStream = function(collection, pipeline, options) {
-  var Collection = require('./collection');
-
-  // Ensure the provided collection is actually a collection
-  if (!(collection instanceof Collection)) {
-    throw new Error(
-      'collection provided to ChangeStream constructor is not an instance of Collection'
+var ChangeStream = function(changeDomain, pipeline, options) {
+  const Collection = require('./collection');
+  const Db = require('./db');
+  const MongoClient = require('./mongo_client');
+
+  this.pipeline = pipeline || [];
+  this.options = options || {};
+  this.cursorNamespace = undefined;
+  this.namespace = {};
+
+  if (changeDomain instanceof Collection) {
+    this.type = CHANGE_DOMAIN_TYPES.COLLECTION;
+    this.serverConfig = changeDomain.s.db.serverConfig;
+
+    this.namespace = {
+      collection: changeDomain.collectionName,
+      database: changeDomain.s.db.databaseName
+    };
+
+    this.cursorNamespace = `${this.namespace.database}.${this.namespace.collection}`;
+  } else if (changeDomain instanceof Db) {
+    this.type = CHANGE_DOMAIN_TYPES.DATABASE;
+    this.namespace = { collection: '', database: changeDomain.databaseName };
+    this.cursorNamespace = this.namespace.database;
+    this.serverConfig = changeDomain.serverConfig;
+  } else if (changeDomain instanceof MongoClient) {
+    this.type = CHANGE_DOMAIN_TYPES.CLUSTER;
+    this.namespace = { collection: '', database: 'admin' };
+    this.cursorNamespace = this.namespace.database;
+    this.serverConfig = changeDomain.topology;
+  } else {
+    throw new TypeError(
+      'changeDomain provided to ChangeStream constructor is not an instance of Collection, Db, or MongoClient'
     );
   }
 
-  var self = this;
-  self.pipeline = pipeline || [];
-  self.options = options || {};
-  self.promiseLibrary = collection.s.promiseLibrary;
-
-  // Extract namespace and serverConfig from the collection
-  self.namespace = {
-    collection: collection.collectionName,
-    database: collection.s.db.databaseName
-  };
+  this.promiseLibrary = changeDomain.s.promiseLibrary;
+  if (!this.options.readPreference && changeDomain.s.readPreference) {
+    this.options.readPreference = changeDomain.s.readPreference;
+  }
 
-  self.serverConfig = collection.s.db.serverConfig;
+  // We need to get the clusterTime as early as possible
+  const isMaster = this.serverConfig.lastIsMaster();
+  if (!isMaster) {
+    throw new Error('ServerConfig does not have an ismaster yet.');
```

Comment:
I know we don't have a particularly hard-set rule about this yet, but I think we should be using `MongoError` here and elsewhere in this file where we are just using `Error`